kotlinx-coroutines-core/common/src/flow/terminal/Reduce.kt

```diff
@@ -84,15 +84,18 @@ public suspend fun <T: Any> Flow<T>.singleOrNull(): T? {
  */
 public suspend fun <T> Flow<T>.first(): T {
     var result: Any? = NULL
+    var expectedException: Throwable? = null
     try {
         collect { value ->
             result = value
-            throw AbortFlowException(NopCollector)
+            val e = AbortFlowException(NopCollector)
+            expectedException = e
+            throw e
         }
     } catch (e: AbortFlowException) {
-        // Do nothing
+        // Do not suppress other cancellation sources
+        if (!e.isSameExceptionAs(expectedException)) throw e
```

Comment:
Now we have two way to check ownership of `AbortFlowException` -- its identity and its `owner` constructor parameter. It seems too much. Can we just use `owner` here instead, just like `combine` and `take` are doing?

Reply:
We can't because at this point we do not have a reference to the "owner" (aka `FlowCollector`).
Alternatively, we can introduce another collect function with the following signature:
`internal suspend inline fun <T> Flow<T>.collectInternal(crossinline action: suspend FlowCollector<T>.(value: T) -> Unit): FlowCollector<T>`, but it may interfere with compiler optimization and obfuscate the code for external readers

Reply:
It still seems safer than checking exception identities. It complicated interaction with various debugging features that copy exceptions. You can do without `collectInternal`, but just create `val flowCollector = object : FlowCollector { ... }` and then call `collect(flowCollector)`. You can also introduce `FlowCollector { ... }` builder function that will go away in Kotlin 1.4 when we turn `FlowCollector` into a `fun interface`.